### PR TITLE
prometheus: Enable live migration of rule configmaps to rule file crds

### PR DESCRIPTION
**Pull request by mxinden**
With this patch the Prometheus Operator checks if there are any
Kubernetes ConfigMaps inside the Prometheus namespace which fullfill the
`Spec.RuleSelector` requirement. If so, it creates a RuleFile for each
key in the `ConfigMap.Spec.Data` map inside the Prometheus namespace.

Tested with a Kubernetes cluster running the kube-prometheus stack :heavy_check_mark:   + additional e2e test.

### CODE REVIEW

**Comment by brancz** [comment]
Happy to offer my prod cluster for giving this a spin :)

**Comment by mxinden** [comment]
@brancz thanks for the feedback. Let me test this on a minikube
kube-prometheus setup first. I don't want to mess up your setup.

File: pkg/prometheus/rulefile.go [hunk 1]
```diff
@@ -158,7 +158,7 @@ func (c *Operator) selectRuleFiles(p *monitoringv1.Prometheus, namespaces []stri
 
 	// sort ruleFiles map
 	filenames := []string{}
-	for k, _ := range ruleFiles {
+	for k := range ruleFiles {
 		filenames = append(filenames, k)
```
[the remark this answers is not in the document]
Comment (mxinden) [reply pkg/prometheus/rulefile.go]: `ruleFiles` is a map.

[the remark this answers is not in the document]
Comment (brancz) [reply pkg/prometheus/rulefile.go]: brain fart, ignore

File: pkg/client/monitoring/v1/types.go [hunk 2]
```diff
@@ -431,7 +431,7 @@ type RuleFile struct {
 // +k8s:openapi-gen=true
 type RuleFileSpec struct {
 	// Content of Prometheus rule file
-	Groups []RuleGroup `json:"groups,omitempty"`
+	Groups []RuleGroup `json:"groups,omitempty" yaml:"groups,omitempty"`
```
[the remark this answers is not in the document]
Comment (mxinden) [reply pkg/client/monitoring/v1/types.go]: Good point. I will switch to the Kuberntes yaml library.

**Comment by mxinden** [comment]
@brancz in regards to: 
> What are your thoughts on possibly annotating configmaps that have been successfully migrated and not attempt to do so again it that’s the case?

I would prefer to leave users config maps untouched. In case they are deploying their setup from a repository, every-time they would redeploy their manifests, they would overwrite the annotation.

Instead I suggest not creating the rule file crd if it already exists:
```golang
_, err := c.mclient.MonitoringV1().RuleFiles(p.Namespace).Create(&ruleFile)
if apierrors.IsAlreadyExists(err) {
	level.Debug(c.logger).Log(
		"msg", "rule file already exists for configmap key",
		"rulefilename", ruleFile.Name,
		"namespace", p.Namespace,
		"prometheus", p.Name,
	)
} else if err != nil {
	return err
}

```

What are your thoughts?

File: pkg/prometheus/rule_configmap_migration.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+			return []monitoringv1.RuleFile{}, errors.Wrapf(
+				err,
+				"unmarshal rules file %v in  configmap '%v' in namespace '%v'",
+				name, cm.Name, cm.Namespace,
+			)
+		}
+
+		ruleFile := monitoringv1.RuleFile{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
```
Comment (brancz) [inline pkg/prometheus/rule_configmap_migration.go]: name needs to be the configmap name + the data key name, as there can be multiple prometheus servers with different rule selectors that have the same files in a configmap

Comment (mxinden) [reply pkg/prometheus/rule_configmap_migration.go]: Great point. :heavy_check_mark:

**Comment by mxinden** [comment]
@brancz @ant31 @community any further thoughts?

**Comment by brancz** [comment]
lgtm

**Pull request merged**